contracts/Uniswap/UniswapAnchoredView.sol

```diff
@@ -0,0 +1,287 @@
+            TokenConfig memory config = getTokenConfigBySymbol(symbols[i]);
+            string memory symbol = symbols[i];
+            bytes32 symbolHash = keccak256(abi.encodePacked(symbol));
+            if (source(messages[i], signatures[i]) != reporter) continue;
+
+            uint reporterPrice = priceData.getPrice(reporter, symbol);
+            uint anchorPrice;
+            if (symbolHash == ethHash) {
+                anchorPrice = ethPrice;
+            } else {
+                anchorPrice = fetchAnchorPrice(config, ethPrice);
+            }
+
+            if (reporterInvalidated == true) {
+                prices[symbolHash] = anchorPrice;
+                emit PriceUpdated(symbol, anchorPrice);
+            } else if (isWithinAnchor(reporterPrice, anchorPrice)) {
+                prices[symbolHash] = reporterPrice;
+                emit PriceUpdated(symbol, reporterPrice);
+            } else {
+                emit PriceGuarded(symbol, reporterPrice, anchorPrice);
+            }
+        }
+    }
+
+    function isWithinAnchor(uint reporterPrice, uint anchorPrice) internal view returns (bool) {
+        if (reporterPrice > 0) {
+            uint anchorRatio = mul(anchorPrice, 100e16) / reporterPrice;
+            return anchorRatio <= upperBoundAnchorRatio && anchorRatio >= lowerBoundAnchorRatio;
+        }
+        return false;
+    }
+
+    /**
+     * @dev Fetches the current token/eth price accumulator from uniswap.
+     */
+    function currentCumulativePrice(TokenConfig memory config) internal view returns (uint) {
+        (uint cumulativePrice0, uint cumulativePrice1,) = UniswapV2OracleLibrary.currentCumulativePrices(config.uniswapMarket);
+        if (config.isUniswapReversed) {
+            return cumulativePrice1;
+        } else {
+            return cumulativePrice0;
+        }
+    }
+
+    /**
+     * @dev Fetches the current eth/usd price from unsiwap, with 6 decimals of precision.
+     *  Conversion factor is 1e18 for eth/usdc market, since we decode uniswap price statically with 18 decimals.
+     */
+    function fetchEthPrice() internal returns (uint) {
+        return fetchAnchorPrice(getTokenConfigBySymbolHash(ethHash), 1e18);
+    }
+
+    /**
+     * @dev Fetches the current token/usd price from uniswap, with 6 decimals of precision.
+     */
+    function fetchAnchorPrice(TokenConfig memory config, uint conversionFactor) internal virtual returns (uint) {
+        (uint nowCumulativePrice, uint oldCumulativePrice, uint oldTimestamp) = pokeWindowValues(config);
+
+        // This should be impossible, but better safe than sorry
+        require(block.timestamp > oldTimestamp, "now must come after before");
+        uint timeElapsed = block.timestamp - oldTimestamp;
+
+        // Calculate uniswap time-weighted average price
+        FixedPoint.uq112x112 memory priceAverage = FixedPoint.uq112x112(uint224((nowCumulativePrice - oldCumulativePrice) / timeElapsed));
```

Comment:
We should add a comment implying that underflow is desired, or note that it's impossible. Otherwise, this should be done via safe-math (maybe that's implied by the `using FixedPoint`?